modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/IgniteH2Indexing.java

```diff
@@ -841,10 +853,35 @@ private ResultSet executeSqlQuery(Connection conn, String sql, Collection<Object
         bindParameters(stmt, params);
 
         try {
+            if (cancel != null) {
+                if (!cancel.compareAndSet(null, new GridAbsClosure() {
+                    @Override public void apply() {
+                        try {
+                            stmt.cancel();
+                            stmt.close();
+                        }
+                        catch (SQLException e) {
+                            throw new IgniteException("Failed to cancel the statement.", e);
+                        }
+                    }
+                }))
+                    throw new GridH2QueryCancelledException();
+            }
+
             return stmt.executeQuery();
         }
         catch (SQLException e) {
+            // Throw special exception.
+            if (e.getMessage().contains("Statement was canceled or the session timed out"))
```

Comment:
Here you have to check for `SQLException.getErrorCode()` but not for message text.